Chat submit metaenter

gjacuna opened the pull request:
Small contribution to chat UX: submit on special enter, ie command+enter on mac or control+enter on pc
No more need to use your mouse to click the button or tab to change focus! just press cmd+enter or control+enter to submit the chat
I'm a newbie, so forgive any mistakes or non-conventions

simonkernel commented on packages/chat/src/views/View.js:
```diff
@@ -112,10 +112,18 @@ const Textarea = ({ state, dispatch }) => {
+const handleMetaEnter = e => {
+  if(e.key === 'Enter' && (e.metaKey || e.ctrlKey)){
+    e.preventDefault()
+    e.target.form.requestSubmit()
```
Thanks for suggesting this feature!

AFAIK, `requestSubmit` is not supported by Safari neither on desktop nor on iOS.

We should find an alternative and maybe either call `submit` on the form or then just call the `sendMessage` fn directly.

Open to other suggestions as long as it works across all major browsers on desktop and mobile.

simonkernel reviewed (changes requested):
Thanks for working on this!

You should run the linter and formatter to fix any styling issues -

```
yarn workspace @kernel/chat lint

yarn workspace @kernel/chat format
```

I also left some comments.

gjacuna commented:
Thanks @simonpure for the feedback. I corrected as per your suggestions and added hot keys to change channels! cmd or ctrl + arrow up or down will iterate over the existing channels
Ran lint and format commands too.

simonkernel commented on packages/chat/src/views/View.js:
```diff
@@ -106,21 +106,38 @@ const Textarea = ({ state, dispatch }) => {
+    await changeChannel(state, dispatch, _channels[_index])
+  } else if (e.key === 'ArrowDown' && (e.metaKey || e.ctrlKey) && state && state.channels) {
+    e.preventDefault()
+    const _channels = Object.values(state.channels)
```
instead of duplicating the logic, let's pull this out into it's own function.

additionally, if we're thinking about the problem we're trying to solve it's really pretending we have an array of channels at some index and we want to either move up or down the index and wrap around on either end.

a little bit of googling brings us to this excellent SO -
https://stackoverflow.com/a/45397704

```
const switchChannel = (state, step) => {
  const { active, channels } = state
  const cnt = Object.keys(channels).length
  const index = Object.values(channels).reduce((acc, {data: { name }, i) => name === active ? i : acc, -1) + step

  // https://stackoverflow.com/a/45397704
  const newIndex = (index % cnt + cnt) % cnt
  const channel = Object.values(channels)[newIndex]
  return channel
}

const STEPS = {'ArrowDown': -1, 'ArrowUp': 1}

const handleKeys = async (state, dispatch, e) => {
  // bail early if meta/ctrl key not pressed
  if (!(e.metaKey || e.ctrlKey)) {
    return
  }

  // switch channels
  if (Object.keys(STEPS).includes(e.key)) {
    const step = STEPS[e.key]
    const channel = switchChannel(state, step)
    dispatch({ type: 'active', payload: channel.data.name })
    dispatch({ type: 'channelId', payload: channel.id })
    return
  }

  // handle other shortcuts like send etc.
}

```

simonkernel replied:
btw I haven't tested this code but you should get the general idea.

simonkernel commented on packages/chat/src/views/View.js:
```diff
@@ -106,21 +106,38 @@ const Textarea = ({ state, dispatch }) => {
+    e.preventDefault()
+    const _channels = Object.values(state.channels)
+    const _current = _channels.find(c => {
+      return state.active === c.data.name
```
fyi in the reduce pattern, you don't change the state directly, but you dispatch an event that changes the state.

simonkernel reviewed (changes requested):
i made some comments to make it a bit more idiomatic.

gjacuna commented:
Hi @simonpure, I just pushed the requested changes. Thanks for the feedback!

simonkernel reviewed (approved):
LGTM
